apps/evm/lib/evm/message_call.ex

```diff
@@ -64,21 +90,31 @@ defmodule EVM.MessageCall do
   end
 
   defp execute(message_call) do
-    # first transitional state
-    message_call = transfer_value_to_recipient(message_call)
-
     message_call
+    |> transfer_funds_if_needed()
     |> prepare_call_execution_env()
     |> execute_call(message_call)
     |> update_state(message_call)
   end
 
+  defp transfer_funds_if_needed(message_call) do
+    if should_transfer_funds?(message_call) do
```

Comment:
This is where we are now checking whether or not we should transfer the funds. The `message_call` structs define a boolean field that informs us as to whether or not we should transfer the value. It's not the greatest implementation (I don't like switching on a boolean), so I'm totally open to other ideas.